File: src/features/settings/model/backup.ts
import { invokeCommand } from "@shared/lib/commands";
import type { AppError } from "@shared/lib/error";
import type { Result } from "@shared/lib/error/result";

export type BackupInfo = {
	filename: string;
	created_at: string;
	size_bytes: number;
};

export const createBackup = async (
	isAuto: boolean,
): Promise<Result<BackupInfo, AppError>> => {
	return invokeCommand<BackupInfo>("create_backup", { isAuto });
};

export const listBackups = async (): Promise<
	Result<BackupInfo[], AppError>
> => {
	return invokeCommand<BackupInfo[]>("list_backups");
};

export const restoreBackup = async (
	backupFilename: string,
): Promise<Result<void, AppError>> => {
	return invokeCommand<void>("restore_backup", { backupFilename });
};

export const deleteBackup = async (
	backupFilename: string,
): Promise<Result<void, AppError>> => {
	return invokeCommand<void>("delete_backup", { backupFilename });
};

export const getBackupDirPath = async (): Promise<Result<string, AppError>> => {
	return invokeCommand<string>("get_backup_dir_path");
};

export const restartApp = async (): Promise<Result<void, AppError>> => {
	return invokeCommand<void>("restart_app");
};

File: src-tauri/src/commands/backup.rs
use tauri::{AppHandle, State};

use crate::db::DbPool;
use crate::error::AppError;
use crate::models::BackupInfo;
use crate::services::backup;

#[tauri::command]
pub async fn create_backup(
    app_handle: AppHandle,
    pool: State<'_, DbPool>,
    is_auto: bool,
) -> Result<BackupInfo, AppError> {
    backup::create_backup(&app_handle, &pool, is_auto).await
}

#[tauri::command]
pub async fn list_backups(app_handle: AppHandle) -> Result<Vec<BackupInfo>, AppError> {
    backup::list_backups(&app_handle)
}

#[tauri::command]
pub async fn restore_backup(
    app_handle: AppHandle,
    backup_filename: String,
) -> Result<(), AppError> {
    backup::restore_backup(&app_handle, &backup_filename)
}

#[tauri::command]
pub async fn delete_backup(app_handle: AppHandle, backup_filename: String) -> Result<(), AppError> {
    backup::delete_backup(&app_handle, &backup_filename)
}

#[tauri::command]
pub async fn get_backup_dir_path(app_handle: AppHandle) -> Result<String, AppError> {
    backup::get_backup_dir_path(&app_handle)
}

#[tauri::command]
pub async fn restart_app(app_handle: AppHandle) -> Result<(), AppError> {
    app_handle.restart();
}
